rust/cardano-chain-follower/src/mithril_turbo_downloader.rs

```diff
@@ -304,52 +305,43 @@ impl MithrilTurboDownloader {
     }
 
     /// Parallel Download, Extract and Dedup the Mithril Archive.
-    async fn dl_and_dedup(&self, location: &str, target_dir: &Path) -> MithrilResult<()> {
+    async fn dl_and_dedup(&self, location: &str) -> anyhow::Result<()> {
         // Get a copy of the inner data to use in the sync download task.
         let inner = self.inner.clone();
-        let location = location.to_owned();
-        let target_dir = target_dir.to_owned();
+
+        debug!("Download Snapshot location='{location}'.");
+        let dl_config = self.inner.cfg.dl_config.clone().unwrap_or_default();
+        let dl_processor =
+            ParallelDownloadProcessor::new(location, dl_config, inner.cfg.chain).await?;
+
+        // Decompress and extract and de-dupe each file in the archive.
+        stats::mithril_extract_started(inner.cfg.chain);
+
+        // We also immediately start downloading now.
+        stats::mithril_dl_started(inner.cfg.chain);
+
+        // Save the DownloadProcessor in the inner struct for use to process the downloaded data.
+        self.inner
+            .dl_handler
+            .set(dl_processor)
+            .map_err(|_| anyhow!("Failed to set the inner dl_handler. Must already be set?"))?;
 
         // This is fully synchronous IO, so do it on a sync thread.
-        let result = spawn_blocking(move || {
+        spawn_blocking(move || {
             stats::start_thread(
                 inner.cfg.chain,
                 stats::thread::name::MITHRIL_DL_DEDUP,
                 false,
             );
-            let result = inner.dl_and_dedup(&location, &target_dir);
+            let result = inner.dl_and_dedup();
             stats::stop_thread(inner.cfg.chain, stats::thread::name::MITHRIL_DL_DEDUP);
             result
         })
-        .await;
-
-        if let Ok(result) = result {
-            return result;
-        }
-
-        stats::mithril_dl_finished(self.inner.cfg.chain, None);
```

Comment:
Actually this is wrong.  Needs to ALWAYS happen when mithril dl finishes or the stats will be corrupted.